feat: configurable path parameters cardinality

lefuturiste opened the pull request:
the idea for this MR was originally discussed in #72 

A change that allow developer to change the cardinality of a given param name for a specific route.

To define a config that goes down the line, we need to use actix `extensions_mut()` method. The external/user developer need to define a middleware that will pass some data down the middleware stack.

Example when defining a route:
```rust
    web::resource("/{source_ids}/{z}/{x}/{y}")
        .wrap_fn(|req, srv| {
            req.extensions_mut().insert::<MetricsConfig>(
                MetricsConfig { cardinality_keep_params: vec!["source_ids".to_string()] }
            );
            srv.call(req)
        })
        .route(web::get().to(get_tile_handler))
```

Feel free to comment or help.

lefuturiste committed:
docs: add a examples with path parameters

the main purpose for me is to have an example on which I can test with
configurable params cardinality feature.

lefuturiste committed:
feat: configurable path cardinality

Allow developer to externally set which on route params they want
to keep in the metrics (as metrics label cardinality) on a
particular route.

lefuturiste commented:
@nlopes You may take a look at this

nlopes commented:
Thanks for sending over. I'll try to take a look sometime this week.

lefuturiste commented:
> Thanks for sending over. I'll try to take a look sometime this week.

Hello, I'm kindly asking if you have some update on this, thanks! 😄

nlopes commented:
Apologies for taking so long. Can you modify your example please? As it stands I don't think it actually exemplifies your change. Or am I misunderstanding it?

lefuturiste commented:
> Apologies for taking so long. Can you modify your example please? As it stands I don't think it actually exemplifies your change. Or am I misunderstanding it?

Yes you are correct, I will update the example.

lefuturiste committed:
docs: clarify params cardinality example

lefuturiste commented:
In theory I should have rebased this fork, and now I added an example. I also need to fix the code so that 404 are not counting as new labels values.

TODO:
- [x] clarify example
- [x] add unit test
- [x] fix 404 behavior
- [x] add test for 404 behavior

lefuturiste committed:
tests: configurable path parameters cardinality and fixes

lefuturiste committed:
fix: discard additional variation of params on 404 responses

lefuturiste commented:
@nlopes can you take a look at this? It should be better now

lefuturiste committed:
docs: introduction to configurable params cardinality

lefuturiste commented:
Hello @nlopes sorry to ask you again, what can be done to get this forward? I know that maintaining opensource software is a burden 😢 .

nlopes commented on README.md:
```diff
@@ -206,3 +206,29 @@ fn main() -> std::io::Result<()> {
+### Configurable routes pattern cardinality
+
+Let's say you have on your app a route to fetch posts by language and by slug `GET /posts/{language}/{slug}`.
+By default, actix-web-prom will provide metrics for the whole route with the label `endpoint` set to the pattern `/posts/{locale}/{slug}`.
```
```suggestion
By default, actix-web-prom will provide metrics for the whole route with the label `endpoint` set to the pattern `/posts/{language}/{slug}`.
```

nlopes commented on src/lib.rs:
```diff
@@ -453,21 +464,26 @@ impl PrometheusMetrics {
         }
 
+        let final_pattern = 
+            if fallback_pattern != pattern && (status == 404 || status == 405) { fallback_pattern }
```
Can you add a comment here on why this is necessary? I'm pretty sure we'll all forget why at some point.

nlopes commented on src/lib.rs:
```diff
@@ -551,10 +567,41 @@ where
+                match strfmt(&full_pattern, &params) {
+                    Ok(mixed_cardinality_pattern) => mixed_cardinality_pattern,
+                    Err(_) => {
+                        // may be we need to log something to warn that we didn't managed to build the pattern?
```
Yes, probably useful to bring in the `log` package.

nlopes reviewed (commented):
Almost there. Apologies, I've forgotten about this.

lefuturiste committed:
refactor: rename pattern as mixed_pattern and add comments

fix: add warn log when failed to build mixed cardinality pattern
docs: typo fix

(apply review suggestions)

lefuturiste commented:
> Almost there. Apologies, I've forgotten about this.

Thanks for your review, I've fix something

lefuturiste committed:
style: apply cargo format
